### Python Pandas Introduction

Here is a short introduction to Pandas using gene expression data. The gene expression data is from the publication and downloaded from  ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE45nnn/GSE45684/suppl/GSE45684_FounderStrainsamples_alignedtoStraingenomes_processedmatrix.txt.gz.
    
Just for conveneience the downloaded file is renamed to a simpler name.

In [2]:
## import pandas
import pandas as pd

#### How to read a tsv file into Pandas Dataframe?

In [4]:
exp_df=pd.read_csv("GSE45684_FounderStrainsamples_exp.txt", sep="\t")
exp_df.head()

/Users/narayan/anaconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Sample  AJ_7103  AJ_7118  AJ_7133  AJ_7163  AJ_7579  AJ_7609  \
0              Strain      A/J      A/J      A/J      A/J      A/J      A/J   
1  ENSMUSG00000000001  6.93507  8.30303  7.02732  8.50000  6.03425  6.70149   
2  ENSMUSG00000000003  0.00000  0.00000  0.00000  0.00000  0.00000  0.00000   
3  ENSMUSG00000000028  0.04168  0.09091  0.06965  0.12500  0.12329  0.08955   
4  ENSMUSG00000000031  0.00000  0.01010  0.00774  0.00000  0.13014  0.00000   

   AJ_7624  AJ_7746  AJ_7761   ...    WSB_11438 WSB_11468 WSB_11483 WSB_11513  \
0      A/J      A/J      A/J   ...      WSB/EiJ   WSB/EiJ   WSB/EiJ   WSB/EiJ   
1  7.09649  7.39416  7.85714   ...      6.61165   6.07377   5.10309   6.32394   
2  0.00000  0.00000  0.00000   ...      0.00000   0.00000   0.00000   0.00000   
3  0.06140  0.05839  0.07143   ...      0.07767   0.04098   0.03093   0.08451   
4  0.00877  0.03650  0.02679   ...      0.03883   0.08197   0.08247   0.07042   

  WSB_11538 WSB_11553 WSB_8772 WSB_8832 WSB_8847 WSB_9149  
0   WSB/EiJ   WSB/EiJ  WSB/EiJ  WSB/EiJ  WSB/EiJ  WSB/EiJ  
1   6.61538   5.75987  6.32099  6.63158  6.70460  5.76316  
2   0.00000   0.00000  0.00000  0.00000  0.00000  0.00000  
3   0.10989   0.02263  0.06173  0.08772  0.02845  0.04386  
4   0.14286   0.38475  0.27160  0.04386  0.18966  0.05263  

[5 rows x 129 columns]

Note that the dataframe has column names and the second row of the data frame is strain name. We can skip the second row while reading the file as dataframe with Pandas. 

#### How to skip a row while readling tsv file into Pandas Dataframe?

In [12]:
exp_df=pd.read_csv("GSE45684_FounderStrainsamples_exp.txt", sep="\t", skiprows=[1])
exp_df.head()

Sample  AJ_7103  AJ_7118  AJ_7133  AJ_7163  AJ_7579  AJ_7609  \
0  ENSMUSG00000000001  6.93507  8.30303  7.02732    8.500  6.03425  6.70149   
1  ENSMUSG00000000003  0.00000  0.00000  0.00000    0.000  0.00000  0.00000   
2  ENSMUSG00000000028  0.04168  0.09091  0.06965    0.125  0.12329  0.08955   
3  ENSMUSG00000000031  0.00000  0.01010  0.00774    0.000  0.13014  0.00000   
4  ENSMUSG00000000037  0.00834  0.00000  0.00000    0.000  0.00000  0.00000   

   AJ_7624  AJ_7746  AJ_7761    ...     WSB_11438  WSB_11468  WSB_11483  \
0  7.09649  7.39416  7.85714    ...       6.61165    6.07377    5.10309   
1  0.00000  0.00000  0.00000    ...       0.00000    0.00000    0.00000   
2  0.06140  0.05839  0.07143    ...       0.07767    0.04098    0.03093   
3  0.00877  0.03650  0.02679    ...       0.03883    0.08197    0.08247   
4  0.00000  0.00000  0.00000    ...       0.00971    0.00820    0.00000   

   WSB_11513  WSB_11538  WSB_11553  WSB_8772  WSB_8832  WSB_8847  WSB_9149  
0    6.32394    6.61538    5.75987   6.32099   6.63158   6.70460   5.76316  
1    0.00000    0.00000    0.00000   0.00000   0.00000   0.00000   0.00000  
2    0.08451    0.10989    0.02263   0.06173   0.08772   0.02845   0.04386  
3    0.07042    0.14286    0.38475   0.27160   0.04386   0.18966   0.05263  
4    0.00000    0.01099    0.00000   0.00000   0.00877   0.00000   0.00000  

[5 rows x 129 columns]

#### How to check the shape or size of a data frame?

In [13]:
exp_df.shape

(24965, 129)

Note that the first column of the dataframe is gene names and it is incorrectly named as Sample. Often we may want to remove a column (or more columns). Let us see how to do that.

#### How to remove one  column (or more columns) from Pandas dataframe?

In [16]:
exp_df.drop(['Sample'],axis = 1).head()

AJ_7103  AJ_7118  AJ_7133  AJ_7163  AJ_7579  AJ_7609  AJ_7624  AJ_7746  \
0  6.93507  8.30303  7.02732    8.500  6.03425  6.70149  7.09649  7.39416   
1  0.00000  0.00000  0.00000    0.000  0.00000  0.00000  0.00000  0.00000   
2  0.04168  0.09091  0.06965    0.125  0.12329  0.08955  0.06140  0.05839   
3  0.00000  0.01010  0.00774    0.000  0.13014  0.00000  0.00877  0.03650   
4  0.00834  0.00000  0.00000    0.000  0.00000  0.00000  0.00000  0.00000   

   AJ_7761  AJ_7776    ...     WSB_11438  WSB_11468  WSB_11483  WSB_11513  \
0  7.85714  6.98077    ...       6.61165    6.07377    5.10309    6.32394   
1  0.00000  0.00000    ...       0.00000    0.00000    0.00000    0.00000   
2  0.07143  0.08654    ...       0.07767    0.04098    0.03093    0.08451   
3  0.02679  0.00000    ...       0.03883    0.08197    0.08247    0.07042   
4  0.00000  0.00000    ...       0.00971    0.00820    0.00000    0.00000   

   WSB_11538  WSB_11553  WSB_8772  WSB_8832  WSB_8847  WSB_9149  
0    6.61538    5.75987   6.32099   6.63158   6.70460   5.76316  
1    0.00000    0.00000   0.00000   0.00000   0.00000   0.00000  
2    0.10989    0.02263   0.06173   0.08772   0.02845   0.04386  
3    0.14286    0.38475   0.27160   0.04386   0.18966   0.05263  
4    0.01099    0.00000   0.00000   0.00877   0.00000   0.00000  

[5 rows x 128 columns]

Insteadof completely deleting the Sample column, let us make that column as rownames of the dataframe.

#### How to assign a column as row names?

In [8]:
exp_df=exp_df.set_index(['Sample'])
exp_df.head()

AJ_7103  AJ_7118  AJ_7133  AJ_7163  AJ_7579  AJ_7609  \
Sample                                                                     
ENSMUSG00000000001  6.93507  8.30303  7.02732    8.500  6.03425  6.70149   
ENSMUSG00000000003  0.00000  0.00000  0.00000    0.000  0.00000  0.00000   
ENSMUSG00000000028  0.04168  0.09091  0.06965    0.125  0.12329  0.08955   
ENSMUSG00000000031  0.00000  0.01010  0.00774    0.000  0.13014  0.00000   
ENSMUSG00000000037  0.00834  0.00000  0.00000    0.000  0.00000  0.00000   

                    AJ_7624  AJ_7746  AJ_7761  AJ_7776    ...     WSB_11438  \
Sample                                                    ...                 
ENSMUSG00000000001  7.09649  7.39416  7.85714  6.98077    ...       6.61165   
ENSMUSG00000000003  0.00000  0.00000  0.00000  0.00000    ...       0.00000   
ENSMUSG00000000028  0.06140  0.05839  0.07143  0.08654    ...       0.07767   
ENSMUSG00000000031  0.00877  0.03650  0.02679  0.00000    ...       0.03883   
ENSMUSG00000000037  0.00000  0.00000  0.00000  0.00000    ...       0.00971   

                    WSB_11468  WSB_11483  WSB_11513  WSB_11538  WSB_11553  \
Sample                                                                      
ENSMUSG00000000001    6.07377    5.10309    6.32394    6.61538    5.75987   
ENSMUSG00000000003    0.00000    0.00000    0.00000    0.00000    0.00000   
ENSMUSG00000000028    0.04098    0.03093    0.08451    0.10989    0.02263   
ENSMUSG00000000031    0.08197    0.08247    0.07042    0.14286    0.38475   
ENSMUSG00000000037    0.00820    0.00000    0.00000    0.01099    0.00000   

                    WSB_8772  WSB_8832  WSB_8847  WSB_9149  
Sample                                                      
ENSMUSG00000000001   6.32099   6.63158   6.70460   5.76316  
ENSMUSG00000000003   0.00000   0.00000   0.00000   0.00000  
ENSMUSG00000000028   0.06173   0.08772   0.02845   0.04386  
ENSMUSG00000000031   0.27160   0.04386   0.18966   0.05263  
ENSMUSG00000000037   0.00000   0.00877   0.00000   0.00000  

[5 rows x 128 columns]

Note that we have successfully used the column as row names and now the dataframe has 128 columns instead of 129. However, the index or rownames still have a name "Sample". 

#### How to remove name of index in Pandas Dataframe?

In [10]:
exp_df.index.name=None
print exp_df.shape
exp_df.head()

(24965, 128)


AJ_7103  AJ_7118  AJ_7133  AJ_7163  AJ_7579  AJ_7609  \
ENSMUSG00000000001  6.93507  8.30303  7.02732    8.500  6.03425  6.70149   
ENSMUSG00000000003  0.00000  0.00000  0.00000    0.000  0.00000  0.00000   
ENSMUSG00000000028  0.04168  0.09091  0.06965    0.125  0.12329  0.08955   
ENSMUSG00000000031  0.00000  0.01010  0.00774    0.000  0.13014  0.00000   
ENSMUSG00000000037  0.00834  0.00000  0.00000    0.000  0.00000  0.00000   

                    AJ_7624  AJ_7746  AJ_7761  AJ_7776    ...     WSB_11438  \
ENSMUSG00000000001  7.09649  7.39416  7.85714  6.98077    ...       6.61165   
ENSMUSG00000000003  0.00000  0.00000  0.00000  0.00000    ...       0.00000   
ENSMUSG00000000028  0.06140  0.05839  0.07143  0.08654    ...       0.07767   
ENSMUSG00000000031  0.00877  0.03650  0.02679  0.00000    ...       0.03883   
ENSMUSG00000000037  0.00000  0.00000  0.00000  0.00000    ...       0.00971   

                    WSB_11468  WSB_11483  WSB_11513  WSB_11538  WSB_11553  \
ENSMUSG00000000001    6.07377    5.10309    6.32394    6.61538    5.75987   
ENSMUSG00000000003    0.00000    0.00000    0.00000    0.00000    0.00000   
ENSMUSG00000000028    0.04098    0.03093    0.08451    0.10989    0.02263   
ENSMUSG00000000031    0.08197    0.08247    0.07042    0.14286    0.38475   
ENSMUSG00000000037    0.00820    0.00000    0.00000    0.01099    0.00000   

                    WSB_8772  WSB_8832  WSB_8847  WSB_9149  
ENSMUSG00000000001   6.32099   6.63158   6.70460   5.76316  
ENSMUSG00000000003   0.00000   0.00000   0.00000   0.00000  
ENSMUSG00000000028   0.06173   0.08772   0.02845   0.04386  
ENSMUSG00000000031   0.27160   0.04386   0.18966   0.05263  
ENSMUSG00000000037   0.00000   0.00877   0.00000   0.00000  

[5 rows x 128 columns]

Now the data frame is simply a matrix of gene expression with sample names as columns and gene ID as row names. Let us explore the data frame.

#### How to check the summary of a dataframe?

In [17]:
exp_df.describe()

AJ_7103       AJ_7118       AJ_7133       AJ_7163       AJ_7579  \
count  24965.000000  24965.000000  24965.000000  24965.000000  24965.000000   
mean       3.102396      3.115466      2.944177      3.348178      2.588834   
std       45.687684     52.597954     47.490102     44.773055     38.437836   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.041680      0.040400      0.043570      0.041670      0.061640   
75%        1.000000      1.000000      1.000000      1.000000      1.000000   
max     5674.068520   6981.262630   6247.736240   4731.020830   4913.993150   

            AJ_7609       AJ_7624       AJ_7746       AJ_7761       AJ_7776  \
count  24965.000000  24965.000000  24965.000000  24965.000000  24965.000000   
mean       2.838941      2.926788      2.741913      2.719740      2.853626   
std       47.701538     45.682257     60.215263     54.334611     49.585033   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.044780      0.043860      0.046200      0.053570      0.038460   
75%        1.000000      1.000000      1.000000      1.000000      1.000000   
max     6512.686570   5927.219300   8816.693430   7786.571430   6787.701920   

           ...          WSB_11438     WSB_11468     WSB_11483     WSB_11513  \
count      ...       24965.000000  24965.000000  24965.000000  24965.000000   
mean       ...           3.461724      3.183974      3.054007      3.558785   
std        ...          57.698305     53.693170     48.277088     57.496498   
min        ...           0.000000      0.000000      0.000000      0.000000   
25%        ...           0.000000      0.000000      0.000000      0.000000   
50%        ...           0.045440      0.049180      0.051550      0.042250   
75%        ...           1.000000      1.000000      1.000000      1.000000   
max        ...        5236.607380   4849.657790   3922.409070   4955.918450   

          WSB_11538     WSB_11553      WSB_8772      WSB_8832     WSB_8847  \
count  24965.000000  24965.000000  24965.000000  24965.000000  24965.00000   
mean       3.268109      2.947340      3.408118      3.376291      3.10726   
std       49.454692     42.389016     55.839714     62.161938     47.73441   
min        0.000000      0.000000      0.000000      0.000000      0.00000   
25%        0.000000      0.000000      0.000000      0.000000      0.00000   
50%        0.054950      0.056580      0.049380      0.043860      0.04742   
75%        1.000000      1.000000      1.000000      1.000000      1.00000   
max     3885.787250   3321.694240   5144.241610   6203.659650   4167.53191   

           WSB_9149  
count  24965.000000  
mean       3.042053  
std       39.762684  
min        0.000000  
25%        0.000000  
50%        0.043860  
75%        1.000000  
max     2867.184210  

[8 rows x 128 columns]